Add support for PCF8574 and PCF8574A GPIO extenders

jpcornil-git opened the pull request:
On ayab hardware, PCF8574s are located at the same addresses as MCP23008s (0x20/0x21) while PCF8574As use different ones (0x38/0x39).

- **Address discovery**
I2C bus is polled for both sets, the first set with both devices detected is selected.

- **Device discovery**
Because MCP23008 IOCON.0  register is read-only (always return 0) while PCF8574/PCF8574A is a single read/write register (output register), both devices can easily be differentiated.
Only side-effect of the discovery for PCF8574 /PCF8574A devices is a spurious 0x01 output.
Except for their I2C addresses, PCF8574  and PCF8574 behave the same

A new GpioExpander base class has been created and MCP23008(modified)/PCF8574 (new) classes are derived from it.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Automatic detection and support for multiple GPIO expander types; solenoids use discovered expanders.
  - I2C device-level detect/read/write support for improved device communication.

- **Improvements**
  - Reduced I2C traffic via output-latch caching and unified expander updates.
  - Hardware pin assignments (encoder, hall sensors, piezo, LEDs) now use Shield constants.
  - Improved hall-sensor wiring detection and initial sensor readings.

- **Refactor**
  - Legacy pin macros replaced by a namespaced Shield API and centralized I2C address table.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jpcornil-git committed:
Add support for PCF8574 GPIO expander

jpcornil-git committed:
Adding two missing 'override' identifiers

jpcornil-git committed:
Fix incorrect i2c_address indexing in Solenoid's constructor

jpcornil-git committed:
Remove redudant definitions for PCF8574 GPIO expander

It shares the same I2C addresses as MCP23008 and the
detections code in solenoid.cpp differentiates both of them.

jpcornil-git commented on lib/devices/gpio_expander/gpio_expander.h:
```diff
@@ -0,0 +1,20 @@
+
+  // Virtual method for derived classes to implement
+  virtual void write(uint8_t address, uint8_t value) = 0;
+
```
Agree, not really required here (base and derived classes never deleted) but it doesn't cost anything and make it clean
=> I'll address in the next commit

jpcornil-git commented on lib/devices/mcp23008/mcp23008.h:
```diff
@@ -2,23 +2,19 @@
 #define MCP23008_OLAT 0x0a
 
-class Mcp23008 {
+class Mcp23008 : GpioExpander {
```
An oversight on my part indeed
=> I'll address in the next commit

jpcornil-git commented on lib/devices/pcf8574/pcf8574.h:
```diff
@@ -0,0 +1,16 @@
+
+  // Write Pcf8574 output register (address is ignored)
+  void write(uint8_t address, uint8_t value) override;
+};
```
Same as above => I'll address in the next commit

jpcornil-git commented on lib/platform/arch/atmelavr/i2c/i2c.cpp:
```diff
@@ -21,6 +21,28 @@ namespace hardwareAbstraction {
+  return is_detected;
+}
+
+uint8_t I2c::read(uint8_t device, uint8_t address) {
+  uint8_t value;
+  ::i2c_start((device << 1) | I2C_WRITE);
+  ::i2c_write(address);
+  ::i2c_rep_start((device << 1) | I2C_READ);
+  value = ::i2c_read(true); // Block forever unless I2C_TIMEOUT is defined
+  ::i2c_stop();
+  return value;
+}
```
true here means "last byte", i.e. end transaction with NACK
see https://github.com/felias-fogg/SoftI2CMaster

jpcornil-git replied:
We learn from each other's :-)

jpcornil-git commented on lib/platform/arch/renesas-ra/i2c/i2c.cpp:
```diff
@@ -5,6 +5,30 @@ namespace hardwareAbstraction {
+  } // FIXME: should signal an error to the desktop app otherwise ?
+  ::Wire.endTransmission();
+  return value;
+}
```
This is a bug indeed, thanks
=> I'll fix in the next commit

jpcornil-git commented on lib/platform/arch/renesas-ra/i2c/i2c.h:
```diff
@@ -10,6 +10,9 @@ class I2c : public I2cInterface {
   I2c();
   ~I2c() = default;
 
+  bool detect(uint8_t device) override;
+  uint8_t read(uint8_t device, uint8_t address) override;
+  void write(uint8_t device, uint8_t value) override;
   void write(uint8_t device, uint8_t address, uint8_t value) override;
```
Same as previous comment => I'll fix in the next commit

jpcornil-git commented on lib/platform/common/hal/hal.h:
```diff
@@ -31,6 +31,9 @@ namespace hardwareAbstraction {
     public:
         virtual ~I2cInterface() = default;
 
+        virtual bool detect(uint8_t device) = 0;
+        virtual uint8_t read(uint8_t device, uint8_t address) = 0;
+        virtual void write(uint8_t device, uint8_t value) = 0;   
         virtual void write(uint8_t device, uint8_t address, uint8_t value) = 0;   
```
i2c_start return a bool already (true = addressed slave ACKed)
see https://github.com/felias-fogg/SoftI2CMaster

jpcornil-git replied:
Yes, this has been fixed now, this was a real one !

jpcornil-git committed:
Fix renesas-ra i2c read method

Address coderabbitai comments:
- fix I2C read method (was using address instead of device in requestFrom)
- make Pcf8574 and Mcp23008 inherit publicly from GpioExpander
- make GpioExpander destructor virtual
- remove const from i2c address parameters (not needed)

VIPQualityPost commented:
Looks good!
I actually had the same nitpick as coderabbit - surprised to see how thorough it can be about semantics.
I would like to keep the PCF8754 address defines, it makes it easier to understand that the addresses are the same, I think.

jpcornil-git commented:
Great ! Did you test the 3 options, PCF8574/PCF8574A as well as MCP23008 (make sure we didn't break anything) ?
I won't be able to access a machine before à few days to test this as well as R4.

Wrt i2c adresses, I initially had 3 sets but given that two were identical I removed one to avoid an unnecessary loop iteration.

Another option would be to remove these defines and just have a list of raw i2caddress to scan given that device detection is done by the code anyway.

We could also remove that list completely and scan the 128 addresses but if someone/in the future we connect something else to the bus we have to make the detection mechanism more selective  -> I would do the above/remove #define for i2c addresses

VIPQualityPost commented:
Sorry, I didn't test at the time of my comment but I have time today to test it at least with both PCF variant. I don't have any shield with MCP expander right now but I do have some parts in the mail, so maybe by the end of the weekend I can try it?

I agree it is a bit wasteful to try for both addresses if they are the same.
Maybe the name could be DEFAULT_I2CEXPANDER_ADDR0/1 rather than mention the model specifically?

jpcornil-git commented:
Thanks :-)

Wrt i2c addresses, I believe that we should just do the following/move everything within shield.h

``` c++
static  const uint8_t gpio_expander_i2c_addresses[][2] = {
    {0x20, 0x21}, // MCP23008 and PCF8574
    {0x38, 0x39}, // PCF8574A
    {0, 0}  // End of list
  };
```

I think also that moving forward we should do all hardware detection at toplevel and e.g. pass `GpioExpander *_gpio_expander[2]` instead of `uint8_t i2cAddress[][2]` to Solenoids' constructor (`Solenoids` doesn't care about i2c, it just want two objects with a `write` method.

VIPQualityPost commented:
Yes, that seems like a very nice solution.

jpcornil-git commented:
I'll modify the code along these lines (over the week-end probably).

jpcornil-git committed:
Refactor Knitter and Solenoids classes

- move GPIO expander detection logic from Solenoids to Knitter class
  - move i2c address definitions from Knitter.cpp to shield.h
  - Solenoids class is now unaware of the GPIO expander type
- use Shield namespace for pin definitions instead of #define
  - Enable type checking and introduce a hierarchical structure for the shield hardware
- change GpioExpander write() method to update() method (no register address)
  - write() is now a derived class method (whenever applicable)

jpcornil-git commented on lib/components/knitter/knitter.cpp:
```diff
@@ -118,6 +124,30 @@ void Knitter::schedule() {
   }
 }
 
+void Knitter::_detectGpioExpanders(hardwareAbstraction::HalInterface *hal, const uint8_t i2cAddress[][2], GpioExpander* gpio_expander[2]) {
+  // FIXME: First one is selected when none are detected -> should raise an error towards desktop app instead
+  int i2C_address_set = 0;
+  for (int id = 0; (i2cAddress[id][0] != 0) || (i2cAddress[id][1] != 0); id++) {
+    if (hal->i2c->detect(i2cAddress[id][0]) && hal->i2c->detect(i2cAddress[id][1])) {
+      i2C_address_set = id;
+      break;
+    };
+  }
+
+  for (int i = 0; i < 2; i++) {
+    // Detect GPIO expander type
+    // MCP23008 IOCON.0 always reads as 0 while PCF8574 will latch the last written value
+    hal->i2c->write(i2cAddress[i2C_address_set][i], MCP23008_IOCON, 0x01);
+    if ((hal->i2c->read(i2cAddress[i2C_address_set][i], MCP23008_IOCON) & 0x01) == 0x00) {
+      Mcp23008 *mcp23008 = new Mcp23008(hal, i2cAddress[i2C_address_set][i]);  
+      mcp23008->write(MCP23008_IODIR, 0);  // Configure as output
+      gpio_expander[i] = (GpioExpander *) mcp23008;
+    } else {
+      gpio_expander[i] = (GpioExpander *) new Pcf8574(hal, i2cAddress[i2C_address_set][i]);
+    }
+  }
+}
```
Solenoid drivers can't be disabled with this hardware and the glitch is not an issue as such.
The Solenoids constructor also calls its reset method to set outputs to their correct reset state.

jpcornil-git commented on lib/components/knitter/knitter.cpp:
```diff
@@ -118,6 +124,30 @@ void Knitter::schedule() {
   }
 }
 
+void Knitter::_detectGpioExpanders(hardwareAbstraction::HalInterface *hal, const uint8_t i2cAddress[][2], GpioExpander* gpio_expander[2]) {
+  // FIXME: First one is selected when none are detected -> should raise an error towards desktop app instead
+  int i2C_address_set = 0;
+  for (int id = 0; (i2cAddress[id][0] != 0) || (i2cAddress[id][1] != 0); id++) {
+    if (hal->i2c->detect(i2cAddress[id][0]) && hal->i2c->detect(i2cAddress[id][1])) {
+      i2C_address_set = id;
+      break;
+    };
+  }
```
Correct but there is no good way to handle this today (therefore the FIXME comment) but i2C_address_set=0 will cause the first entry to be used which is the same as of today when only mcp23008 devices are expected/supported.

jpcornil-git commented on lib/components/knitter/knitter.h:
```diff
@@ -48,6 +48,8 @@ class Knitter : protected API {
   void schedule();
 
  private:
+  // Detect addresses and type of both GPIO expanders
+  void _detectGpioExpanders(hardwareAbstraction::HalInterface *hal, const uint8_t i2cAddress[][2], GpioExpander* gpio_expander[2]);
```
See above/previous comment, it is preferable to mimic the behavior of the legacy implementation as nothing can be reported back to the desktop application for the time being.

jpcornil-git commented on lib/components/solenoids/solenoids.cpp:
```diff
@@ -34,7 +32,7 @@ void Solenoids::set(uint8_t solenoid, bool state) {
   uint16_t values = _states;
   for (int i = 0; i < 2; i++) {
-    _mcp23008[i]->write(MCP23008_OLAT, (uint8_t)(values & 0xff));
+    _gpio_expander[i]->update((uint8_t)(values & 0xff));
```
This will never be the case in today's code given the current implementation (first devices in the list used when nothing is found) but this should be done when initialization failure will be properly handled/reported to the desktop app.

jpcornil-git replied:
I'll implement it anyway ... it doesn't harm really

jpcornil-git commented on lib/devices/pcf8574/pcf8574.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "pcf8574.h"
+
+void Pcf8574::update(uint8_t value) {
+  if ((value == _output_latch_cache) && (!_cache_invalid)) {
+    return;
+  }
+  _output_latch_cache = value;
+  _cache_invalid = false;
+
+  _hal->i2c->write(_i2cAddress, value);
+}
```
This is a good point indeed and should be done once write() return OK/NOK for all supported platforms (it doesn't return anything today)

jpcornil-git replied:
Yes please

jpcornil-git commented on lib/devices/pcf8574/pcf8574.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "pcf8574.h"
+
+void Pcf8574::update(uint8_t value) {
+  if ((value == _output_latch_cache) && (!_cache_invalid)) {
+    return;
+  }
```
This is the case already, see gpio_expander.h (base class)

jpcornil-git committed:
Code improvements and refactoring for better maintainability and review.

jpcornil-git commented:
@VIPQualityPost, I think I'm done with this one => let me know how your tests with the 3 GPIO expansion flavors go!

jpcornil-git commented on lib/devices/mcp23008/mcp23008.cpp:
```diff
@@ -1,19 +1,17 @@
+  if (reg == MCP23008_OLAT) {
+    if ((value == _output_latch_cache) && (!_cache_invalid)) {
       return;
     }
-    _olat_cache = value;
-    _olat_cache_invalid = false;
+    _output_latch_cache = value;
+    _cache_invalid = false; // TODO: invalidate after write() failure when supported
   }
 
-  _hal->i2c->write(_i2cAddress, address, value);
+  _hal->i2c->write(_i2cAddress, reg, value);
```
_cache_invalid is initialized to true in the base class constructor and therefore the first call to Mcp23008 or Pcf8574 update() method (calling write(MCP23008_OLAT, ...) for mcp23008) will always skip the if statement and initialize _output_latch_cache  correctly.

There is no correct value for _output_latch_cache  unless you read the hardware value from the device (not generic/possible for e.g. pcf8574)

jpcornil-git replied:
Yep and you created #5 to track this

jpcornil-git commented on lib/components/hallsensor/hallsensor.cpp:
```diff
@@ -3,14 +3,34 @@
+  // Default pin mode (analog/digital for pin1/2 respectively)
+  _hal->pinMode(_sensorPin1, INPUT);
+  _hal->pinMode(_sensorPin2, INPUT_PULLUP);
+
```
Agree, I'll fix

VIPQualityPost commented:
My MCP hardware finally arrives, so I can do some testing this weekend... sorry for the delay!

jpcornil-git commented on lib/components/hallsensor/hallsensor.cpp:
```diff
@@ -165,5 +181,20 @@ bool HallSensor::_detectCarriage() {
+    // Analog sensor, read voltage
+    _sensorValue = _hal->analogRead(_sensorPin1);
+  }
+}
```
- `_config` is always defined (set together with `_machine` in `_apiRequestInit` and `_readSensor` is only called when `_machine` is defined).
- For digital sensor, K is driven LOW while L is driven HIGH 
- HALLSENSOR_DIGITAL should never be defined when "_sensorPin2 == PIN_NONE", i.e. this is a configuration error

jpcornil-git commented on lib/platform/common/shield/shield.h:
```diff
@@ -1,19 +1,47 @@
+
+    // Hall detectors
+    struct HallDetectors {
+        // Analog
+        struct Analog {
+            static constexpr uint8_t EOL_R_PIN = 0;  // Right
+            static constexpr uint8_t EOL_L_PIN = 1;  // Left
+        };
+        // Digital (KH910 RHS)
+        struct Digital {
+            static constexpr uint8_t EOL_R_L_PIN = 7;
+            static constexpr uint8_t EOL_R_DETECT_PIN = 8;
+        };
```
`EOL_R_PIN` can either be analog  or digital based on  `HALLSENSOR_DIGITAL` value and, based on that,  `_readSensor` is using `analogRead` or `digitalRead` respectively

jpcornil-git committed:
Fix hallsensor's pin mode for the second pin

- Restore INPUT mode for the second pin of the hallsensor when the hardware fix is not detected
- Change mcp23008 #defines to use class constants

VIPQualityPost commented:
The I2C devices are all working as expected for me with both PCF variant and MCP!

jpcornil-git commented:
@VIPQualityPost If not done already, could you also retest with the latest commit (0c3ac31) to make sure I didn't break anything with it ?
You may also want to test (on one hardware) EOL_R on KH910 with the @jonathanperret hardware fix as it is integrated as well.

Once the two PRs I created earlier today to support PCF8574 and EOL_R on KH910 on ayab-firmware/AyabAsync will be merged, it should match behavior of this one.

VIPQualityPost commented:
I do'nt have KH910, maybe someone else can test with that? I only have KH940 and KH270.

jonathanperret commented:
> I do'nt have KH910, maybe someone else can test with that? I only have KH940 and KH270.

I can test the 910 next week.

jpcornil-git commented:
> > I do'nt have KH910, maybe someone else can test with that? I only have KH940 and KH270.
> 
> I can test the 910 next week.

Thanks, I'll try to get access to my daughter's machine next week too to do some testing with her Uno R4 (validate i2c modifications for RA4M1) and Mega2560. I can also try to modify her shield with the EOL_R fix ... if I'm allowed to :-)

VIPQualityPost commented:
I forgot I had RA4M1 now, I can test on that too.

jpcornil-git commented:
> I forgot I had RA4M1 now, I can test on that too. 

Thanks !

Have you tried flashing ayab-esp32 and enjoying WiFi connectivity with ayab-desktop? (You can also update the default RA4M1 firmware using the UI). Let me know if you need help with that.

jpcornil-git committed:
Fix _readSensor() to handle _config == nullptr properly

In the object constructor, _readSensor() is called before _config is
initialized and this leads to a nullptr dereference and an undefined
return value. This is also the case when _apiRxIndicateState() is
called before _config is set by an _apiRequestInit() call.

This is not critical but not clean either.

jpcornil-git committed:
Fix the hall sensor pin definitions

Analog inputs were previously defined as analog channels (analogRead
accepts both analog channel and pin numbering) but now that EOL_R_PIN
is also used as a digital input for the KH910, it MUST be defined as
a digital.
